keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md

```diff
@@ -0,0 +1,606 @@
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a kubeadm user that wishes to modify the KubeletConfiguration that kubeadm stores in the cluster
+before joining new nodes, I wish to have easy access to a static location such as `kube-system/kubelet-config`
+without looking up the control plane version first.
+
+#### Story 2
+
+As a kubeadm maintainer, I wish to simplify the logic in kubeadm that manages the kubelet configuration
+read/write operations.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+NONE
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Risk 1
+
+**Users do not have sufficient visibility of the change**
+
+Communicate the change on the following channels to ensure as much visibility as possible:
+- kubeadm release notes
+- kubernetes-dev mailing list
+- #kubeadm on k8s slack
+- Blog posts, Twitter, Reddit, other
+
+#### Risk 2
+
+**Users that do not use `kubeadm upgrade` would not be able to join new nodes**
+
+In the Alpha stage of the change, the command `kubeadm upgrade apply` will be used to create
+a ConfigMap and the related RBAC rules that follow the new naming format. Nodes that are joined
+to the cluster later would be able to use this new ConfigMap. During Beta the presence of these
+new API objects will become a requirement, unless the users opt-out by disabling the feature gate
+stored in the `ClusterConfiguration`.
+
+Cluster API is an example of a project that uses kubeadm for cluster creation, but does not use the
+`kubeadm upgrade` command for upgrades. These upgrades are called "immutable node upgrades", where new
+version nodes join the cluster and replace the old ones. Similar projects and users would have to
+implement their own means to support the Alpha->Beta transition of this feature by populating the required
+ConfigMap and RBAC rules before joining new nodes during an immutable upgrade.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
```

Comment:
I think that we need also a using test for spinning up a cluster with the feature gate enabled, but without kubelet skew

Reply:
do you mean adding an extra kubelet skew test - e.g. kubeadm N deploying kubelet N-1?
we can add such a test when the kubelet v1beta2 is added.

i think i should rename **Skew e2e test job**` in the doc to **Join e2e test job**, because i'm realizing the proposed kubeadm vs control-plane skew test case is not important and the important part is that we ensure "join" gets the correct CM.

Reply:
> we can add such a test when the kubelet v1beta2 is added.

added under "future e2e tests"

> i think i should rename Skew e2e test job` in the doc to Join e2e test job

done.